miden-crypto/src/encryption/xchacha/mod.rs

```diff
@@ -0,0 +1,506 @@
+        // is different than the one used in the `chacha20poly1305`. This solution will
+        // no longer be needed once `chacha20poly1305` gets a new release with a version of
+        // the `rand` dependency matching ours
+        use chacha20poly1305::aead::rand_core::SeedableRng;
+        let mut seed = [0_u8; 32];
+        rand::RngCore::fill_bytes(rng, &mut seed);
+        let rng = rand_hc::Hc128Rng::from_seed(seed);
+
+        Nonce {
+            inner: XChaCha20Poly1305::generate_nonce(rng),
+        }
+    }
+
+    /// Creates a new nonce from the provided array of bytes
+    pub fn from_slice(bytes: &[u8; NONCE_SIZE_BYTES]) -> Self {
+        Nonce { inner: (*bytes).into() }
+    }
+}
+
+/// A 256-bit secret key
+#[derive(Debug, PartialEq, Eq)]
+pub struct SecretKey([u8; SK_SIZE_BYTES]);
+
+impl SecretKey {
+    /// Creates a new random secret key using the default random number generator
+    #[cfg(feature = "std")]
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        let mut rng = rand::rng();
+        Self::with_rng(&mut rng)
+    }
+
+    /// Creates a new random secret key using the provided random number generator
+    pub fn with_rng<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
+        // we use a seedable CSPRNG and seed it with `rng`
+        // this is a work around the fact that the version of the `rand` dependency in our crate
+        // is different than the one used in the `chacha20poly1305`. This solution will
+        // no longer be needed once `chacha20poly1305` gets a new release with a version of
+        // the `rand` dependency matching ours
+        use chacha20poly1305::aead::rand_core::SeedableRng;
+        let mut seed = [0_u8; 32];
+        rand::RngCore::fill_bytes(rng, &mut seed);
+        let rng = rand_hc::Hc128Rng::from_seed(seed);
+
+        let key = XChaCha20Poly1305::generate_key(rng);
+        Self(key.into())
+    }
+
+    /// Encrypts, as Felt-s, and authenticates the provided data using this secret key and a random
+    /// nonce
+    #[cfg(feature = "std")]
+    pub fn encrypt(&self, data: &[Felt]) -> Result<EncryptedFeltData, EncryptionError> {
+        self.encrypt_with_associated_data(data, &[])
+    }
+
+    /// Encrypts, as Felt-s, the provided data and authenticates both the ciphertext as well as
+    /// the provided associated data using this secret key and a random nonce
+    #[cfg(feature = "std")]
+    pub fn encrypt_with_associated_data(
+        &self,
+        data: &[Felt],
+        associated_data: &[Felt],
+    ) -> Result<EncryptedFeltData, EncryptionError> {
+        let mut rng = rand::rng();
+
+        let nonce = Nonce::with_rng(&mut rng);
+        self.encrypt_with_nonce(data, associated_data, nonce)
+    }
+
+    /// Encrypts the provided data using this secret key and a specified nonce
+    pub fn encrypt_with_nonce(
+        &self,
+        data: &[Felt],
```

Comment:
The use case for having this scheme is native encryption, and so we expect the data to be serialized to bytes. I'm not sure we need the variants that accept `data: &[Felt]`